### ISSUE TITLE
process_batch in val.py is Matching Boxes Incorrectly

### ISSUE THREAD

**Issue by Charlie-AuZone** [ultralytics/yolov5#12673]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Validation

### Bug

In some scenarios the process_batch function in val.py produces incorrect matching of detections to labels. 

In the example given, the `box_iou() ` function is copied from https://github.com/ultralytics/yolov5/blob/050c72cbba9ca24dcff103fca54caf21d7e8b527/utils/metrics.py#L266-L285
and the `process_batch()` function is copied from https://github.com/ultralytics/yolov5/blob/050c72cbba9ca24dcff103fca54caf21d7e8b527/val.py#L90-L113

We expect the 2nd prediction (0.25, 0.25, 0.75, 0.75) to be matched with the label (0.25, 0.25, 0.75, 0.75), since the 2nd prediction is the exact same as box and label as the label. However, the process_batch function matches the 1st prediction (0.20, 0.20, 0.70, 0.70) to the label instead, causing the 2nd prediction, which is *exactly* the same as the label, to be considered incorrect.

### Environment

- OS: Windows 10
- Python: Python 3.10.11
- PyTorch: torch==1.10.1
- YOLOv5: Master branch

### Minimal Reproducible Example

```python
import numpy as np
import torch
def box_iou(box1, box2, eps=1e-7):
    # https://github.com/pytorch/vision/blob/master/torchvision/ops/boxes.py
    """
    Return intersection-over-union (Jaccard index) of boxes.

    Both sets of boxes are expected to be in (x1, y1, x2, y2) format.
    Arguments:
        box1 (Tensor[N, 4])
        box2 (Tensor[M, 4])
    Returns:
        iou (Tensor[N, M]): the NxM matrix containing the pairwise
            IoU values for every element in boxes1 and boxes2
    """

    # inter(N,M) = (rb(N,M,2) - lt(N,M,2)).clamp(0).prod(2)
    (a1, a2), (b1, b2) = box1.unsqueeze(1).chunk(2, 2), box2.unsqueeze(0).chunk(2, 2)
    inter = (torch.min(a2, b2) - torch.max(a1, b1)).clamp(0).prod(2)

    # IoU = inter / (area1 + area2 - inter)
    return inter / ((a2 - a1).prod(2) + (b2 - b1).prod(2) - inter + eps)

def process_batch(detections, labels, iouv):
    """
    Return correct prediction matrix.

    Arguments:
        detections (array[N, 6]), x1, y1, x2, y2, conf, class
        labels (array[M, 5]), class, x1, y1, x2, y2
    Returns:
        correct (array[N, 10]), for 10 IoU levels
    """
    correct = np.zeros((detections.shape[0], iouv.shape[0])).astype(bool)
    iou = box_iou(labels[:, 1:], detections[:, :4])
    correct_class = labels[:, 0:1] == detections[:, 5]
    for i in range(len(iouv)):
        x = torch.where((iou >= iouv[i]) & correct_class)  # IoU > threshold and classes match
        if x[0].shape[0]:
            matches = torch.cat((torch.stack(x, 1), iou[x[0], x[1]][:, None]), 1).cpu().numpy()  # [label, detect, iou]
            if x[0].shape[0] > 1:
                matches = matches[matches[:, 2].argsort()[::-1]]
                matches = matches[np.unique(matches[:, 1], return_index=True)[1]]
                # matches = matches[matches[:, 2].argsort()[::-1]]
                matches = matches[np.unique(matches[:, 0], return_index=True)[1]]
            correct[matches[:, 1].astype(int), i] = True
    return torch.tensor(correct, dtype=torch.bool, device=iouv.device)
    
if __name__== "__main__":
    iou = 0.5

    # detections    (array[N, 6]),      x1, y1, x2, y2, conf, class
    detections = np.array([
        [0.20, 0.20, 0.70, 0.70, 1. , 1. ],
        [0.25, 0.25, 0.75, 0.75, 1. , 1. ],
    ])

    # labels        (array[M, 5]),      class, x1, y1, x2, y2
    ground_truth = np.array([
        [1. , 0.25, 0.25, 0.75, 0.75],
    ])

    detections = torch.from_numpy(detections)
    ground_truth = torch.from_numpy(ground_truth)
    
    yolo_matched = process_batch(detections, ground_truth, torch.Tensor([iou])).flatten().tolist()
    print(yolo_matched)

    # This assert fails
    assert yolo_matched == [False, True]
        
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Charlie-AuZone hello! Thank you for the detailed issue report. It seems like there might be an issue with the matching logic in `process_batch` that's causing the incorrect matching of detections to labels.

The matching process should indeed prioritize the highest IoU for each detection-label pair, and in the case of a perfect match (IoU of 1.0), it should not be overridden by a lower IoU match.

I'll look into this and address the bug. In the meantime, if you have any insights or suggestions on how to improve the matching logic, feel free to share them or contribute directly with a PR. Your efforts in helping improve YOLOv5 are greatly appreciated! 🙌

Keep an eye on the issue tracker for updates on this bug. Thanks again for bringing this to our attention!
